server/README.md

```diff
@@ -0,0 +1,236 @@
+
+## Data concurrency
+
+The server needs to handle multiple queries concurrently.
+
+For instance if two concurrent requests to create an invitation to the same person
+should have the same outcome than if those requests were consecutive (i.e. the
+second request should detect an invitation already exists and return it).
+
+## Data causality
+
+Given changes in the data come from different client, the server is responsible
+to guarantee causality on all data.
+
+### Example 1
+
+- At T1 Alice query the server for all the certificates
+- At T2 a new certificates from Bob reaches the server with timestamp T0 < T1 (Bob's
+  clock might be off, or the query took a long time to reach the server)
+
+In this case, if Alice got a certificate with a timestamp more recent that Bob's certificate,
+then accepting Bob's certificate would break causality !
+
+### Example 2
+
+- At T1 the server receive block create from Alice in realm R1
+- At T2 the server receive a certificate unsharing with Alice realm R1 with timestamp T0
+
+In this case, and considering the server has accepted the unsharing certificate,
+any client accessing the block before T2 will consider everything is fine, while
+any client accessing the block after T2 will reject it given it appears to have
+been created after Alice lost her access to the realm !
+
+## Causality & concurrency model
+
+From the server point of view, there is basically four different kinds of data:
+
+- Certificates
+- Blocks
+- Vlobs
+- Invitations
+
+### Certificates
+
+Any change done by a client that should be visible by the server is provided as a
+certificate (i.e. a certificate is an unencrypted document signed by the client's device).
+
+A very simple way to ensure causality and concurrency on certificates would be:
+
+- Have a single big lock in the server that must be taken before any attempt to
+  add a certificate (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  (handles causality).
+
+However this is pretty wasteful given most certificates are unrelated (e.g. certif A
+add a new user, certif B does a key rotation in a realm). So instead we reduce the
+size of the lock by introducing the concept of topic.
+
+Each type of certificate is part of one of the following topics:
+
+- `common`
+- `sequester`, which depends on `common`
+- `realm`, which depends on `common` and uses the realm ID as discriminant
+- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant
+
+So the actual rule used in the server is:
+
+- Have lock in the server for each couple topic + discriminant
+- Take the lock for the topic + discriminant couple the certificate we want to add
+  belongs to AND to other topics our topic may depend on (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  in the topics we have locked (handles causality).
```

Comment:
Also, the section does not mention the causality dependencies that exist between certificates and other types of data such as blocks or vlobs.